API changes for cluster autoscaling options

elmiko opened the pull request:
This change adds options for the cluster autoscaler to the api doc.
These additions mirror the options that are currently exposed on
OpenShift through the ClusterAutoscaler resource with the addition of
the `ExtraArgs` field to allow arbitrary flags on the autoscaler.

Also cleanup the references to "Autoscaler" and "Autoscaling" to be
inline with upstream capitalization.

This PR is in support of [HOSTEDCP-82](https://issues.redhat.com/browse/HOSTEDCP-82)

enxebre commented on docs/api.md:
```diff
@@ -84,8 +84,96 @@ type AddonsSpec struct {
+	SkipNodesWithLocalStorage *bool `json:"skipNodesWithLocalStorage,omitempty"`
+
+    // Allow users to specify additional arguments that should be passed to the cluster-autoscaler.
+    ExtraArgs string `json:"ExtraArgs,omitempty"`
```
I'd be inclined to not expose ExtraArgs and have a just a predictable gated list.

elmiko replied:
i don't mind removing it if folks object, but my rationale for inclusion is that having a field like this makes it easier to inject commands to the autoscaler for testing and debugging purposes.

enxebre commented:
This is early in the discussion as we haven't decided on addons but I see autoscaling as something which is always active. You opt-in via particular nodePools. So I see these settings belonging to something like `hostedCluster.spec.Autoscaling` rather than to an add-on.

elmiko commented:
> This is early in the discussion as we haven't decided on addons yet but I see autoscaling as something which is always active. You opt-in via particular nodePools. So I see these settings belonging to something like `hostedCluster.spec.Autoscaling` rather than to an add-on.

+1, @derekwaynecarr made a similar statement in slack. i'm happy to move this into `hostedClsuter.spec.Autoscaling`.

elmiko commented:
* removed ExtraArgs param
* added `Autoscaling` as a top level field in `HostedClusterSpec`
* renamed `AutoscalerAddonSpec` to `ClusterAutoscaling`, inspired by the `Networking` field

derekwaynecarr commented on docs/api.md:
```diff
@@ -118,8 +120,93 @@ type AddonsSpec struct {
+
+	// Minimum and maximum number of different GPUs in cluster, in the format <gpu_type>:<min>:<max>.
+	// Cluster autoscaler will not scale the cluster beyond these numbers. Can be passed multiple times.
+	GPUS []GPULimit `json:"gpus,omitempty"`
```
i would omit this for the moment.

ironcladlou commented:
Do we anticipate autoscaling of control plane components? If so, will its configuration be exposed to the user? And at what component granularity? Just something I thought of given the name is HostedCluster.AutoScaling which doesn't indicate right away that it's about node scaling.

ironcladlou commented on docs/api.md:
```diff
@@ -118,8 +120,93 @@ type AddonsSpec struct {
+
+	// How long after scale up that scale down evaluation resumes
+	// +kubebuilder:validation:Pattern=([0-9]*(\.[0-9]*)?[a-z]+)+
+	DelayAfterAdd *string `json:"delayAfterAdd,omitempty"`
```
What is the actual type of this field? Is it a `time.Duration` or some string enum? If so where are the valid values enumerated?

ironcladlou commented on docs/api.md:
```diff
@@ -118,8 +120,93 @@ type AddonsSpec struct {
+
+	// How long after node deletion that scale down evaluation resumes, defaults to scan-interval
+	// +kubebuilder:validation:Pattern=([0-9]*(\.[0-9]*)?[a-z]+)+
+	DelayAfterDelete *string `json:"delayAfterDelete,omitempty"`
```
As above re: the type of this field

ironcladlou commented on docs/api.md:
```diff
@@ -118,8 +120,93 @@ type AddonsSpec struct {
+
+	// How long after scale down failure that scale down evaluation resumes
+	// +kubebuilder:validation:Pattern=([0-9]*(\.[0-9]*)?[a-z]+)+
+	DelayAfterFailure *string `json:"delayAfterFailure,omitempty"`
```
As above re: the type of this field

ironcladlou commented on docs/api.md:
```diff
@@ -118,8 +120,93 @@ type AddonsSpec struct {
+
+	// How long a node should be unneeded before it is eligible for scale down
+	// +kubebuilder:validation:Pattern=([0-9]*(\.[0-9]*)?[a-z]+)+
+	UnneededTime *string `json:"unneededTime,omitempty"`
```
As above re: the type of this field

ironcladlou commented on docs/api.md:
```diff
@@ -118,8 +120,93 @@ type AddonsSpec struct {
+
+type ScaleDownConfig struct {
+	// Should CA scale down the cluster
+	Enabled bool `json:"enabled"`
```
The scale down config itself is an optional object. If an omitted scale down config is interpreted as "scale down disabled" (I'm assuming — this needs to be specified), what is the purpose of specifying the field but declaring it enabled=false vs. omitting the field? Would it be possible to remove this field entirely?

ironcladlou commented on docs/api.md:
```diff
@@ -118,8 +120,93 @@ type AddonsSpec struct {
+	// Maximum number of nodes in all node groups.
+	// Cluster autoscaler will not grow the cluster beyond this number.
+	// +kubebuilder:validation:Minimum=0
+	MaxNodesTotal *int32 `json:"maxNodesTotal,omitempty"`
```
What's the default?

elmiko replied:
in the update i've changed this to force the user to specify. i'm not sure what the best option for default max size is, but it seemed like the most important piece of the autoscaling to force a user into specifying.

ironcladlou commented on docs/api.md:
```diff
@@ -118,8 +120,93 @@ type AddonsSpec struct {
+
+	// Minimum and maximum number of cores in cluster, in the format <min>:<max>.
+	// Cluster autoscaler will not scale the cluster beyond these numbers.
+	Cores *ResourceRange `json:"cores,omitempty"`
```
What's the default?

ironcladlou commented on docs/api.md:
```diff
@@ -118,8 +120,93 @@ type AddonsSpec struct {
+
+	// Minimum and maximum number of gigabytes of memory in cluster, in the format <min>:<max>.
+	// Cluster autoscaler will not scale the cluster beyond these numbers.
+	Memory *ResourceRange `json:"memory,omitempty"`
```
What's the default?

ironcladlou commented on docs/api.md:
```diff
@@ -118,8 +120,93 @@ type AddonsSpec struct {
+	// Constraints of autoscaling resources
+	ResourceLimits *ResourceLimits `json:"resourceLimits,omitempty"`
+
+	// Configuration of scale down operation
```
What's he default behavior in the absence of this field?

ironcladlou commented on docs/api.md:
```diff
@@ -118,8 +120,93 @@ type AddonsSpec struct {
 // TODO maybe we have profiles for scaling behaviors
-type AutoScalerAddonSpec struct {
-    ResourceLimits LimitRange
+type ClusterAutoscaling struct {
```
General note: bool fields in API are generally suspicious, and I also wonder how many of these are often set by users and which might be considered esoteric and could be omitted from the minimum viable first draft of the API.

elmiko replied:
i've left it in for now, but i think in the specific case of the `ScaleDownEnabled` (new name added in recent push), we could probably evaluate just dropping it completely.

i am curious to hear more about your thoughts behind 
> bool fields in API are generally suspicious
we use this bool in the CAO, and if it's bad form i'd like to evaluate if we can make this easier there as well.

elmiko commented:
> Do we anticipate autoscaling of control plane components? If so, will its configuration be exposed to the user? And at what component granularity? Just something I thought of given the name is HostedCluster.AutoScaling which doesn't indicate right away that it's about node scaling.

just to add some details from our meeting today.
* i think we are in agreement that there will not be autoscaling applied to the control plane
* we didn't come to a decision about the name, i will leave it as `Autoscaling` for now but happy to take suggestions (i mentioned `ComputeNodeAutoscaling` in the meeting as a possible alternate)
* dropping the max resource fields for CPU, Mem, and GPU. these will be set based on needs of openshift dedicated initially.
* dropping excess scale down fields
* change fields from pointer to concrete types and add defaults

elmiko commented:
updated the api to remove some of the resource stuff and also the scale down section. i tried to focus on fields that an operator might want to tune for a specific cluster. i left the pointer references because there are cases where a zero value could be misinterpreted, testing for nil is what we have done in the CAO. i also noted all the defaults in the comments.

i think we could probably drop the `ScaleDownEnabled`, it should probably be always enabled.

i'm also not sure if we need `BalanceSimilarNodeGroups`, but left it in for now.

ironcladlou commented on docs/api.md:
```diff
@@ -118,8 +121,46 @@ type AddonsSpec struct {
+
+	// Should CA scale down the cluster
+    // default: true
+	ScaleDownEnabled bool `json:"enabled,omitempty"`
```
I propose we remove this field and have a followup discussion about whether there's some implied policy object to drive things like turning the entire system on or off, etc.

elmiko replied:
that makes sense to me. i think we could easily just drop this and have scale down always enabled, and happy to have the followup conversation.

ironcladlou replied:
Sounds like we agreed to remove this but it remains

elmiko replied:
i just haven't pushed a new version yet. was waiting to gather a few more responses.

ironcladlou commented:
Okay, with fresh eyes there's something I don't understand: How exactly does one select which NodePools participate in autoscaling?

elmiko commented:
> Okay, with fresh eyes there's something I don't understand: How exactly does one select which NodePools participate in autoscaling?

as it stands today, the autoscaler will only operate on `MachineSets` and `MachineDeployments`. to have the autoscaler recognize these resources for scaling they must be annotated with the folliwing:
* `machine.openshift.io/cluster-api-autoscaler-node-group-min-size`
* `machine.openshift.io/cluster-api-autoscaler-node-group-max-size`

we gate this through the `MachineAutoscaler` resource in openshift.

if the autoscaler is going to need to understand `NodePool` resources instead of `MachineSets` or `MachineDeployments`, then we will need to adjust the autoscaler implementation but i would imagine the method of inclusion would stay relatively similar (unless we really didn't want it).

ironcladlou commented:
> > Okay, with fresh eyes there's something I don't understand: How exactly does one select which NodePools participate in autoscaling?
> 
> as it stands today, the autoscaler will only operate on `MachineSets` and `MachineDeployments`. to have the autoscaler recognize these resources for scaling they must be annotated with the folliwing:
> 
> * `machine.openshift.io/cluster-api-autoscaler-node-group-min-size`
> * `machine.openshift.io/cluster-api-autoscaler-node-group-max-size`
> 
> we gate this through the `MachineAutoscaler` resource in openshift.
> 
> if the autoscaler is going to need to understand `NodePool` resources instead of `MachineSets` or `MachineDeployments`, then we will need to adjust the autoscaler implementation but i would imagine the method of inclusion would stay relatively similar (unless we really didn't want it).

Thanks for that context. That seems useful for our implementation, but I think what I meant to ask was how does an end user of hypershift select which NodePools participate in autoscaling? Or are we saying that all NodePools owned by a HostedCluster are (at least in the 1st rev) subject to the rules defined in the HostedCluster Autoscaling configuration defined in this PR?

enxebre commented:
>Thanks for that context. That seems useful for our implementation, but I think what I meant to ask was how does an end user of hypershift select which NodePools participate in autoscaling? Or are we saying that all NodePools owned by a HostedCluster are (at least in the 1st rev) subject to the rules defined in the HostedCluster Autoscaling configuration defined in this PR?

Autoscaling is already running and working in the main branch. You can opt-in for a nodePool to be autoscaled by setting `spec.Autoscalin.min` and `spec.Autoscalin.max`

This PR is just exposing some tweakable flags for the cluster-autoscaler binary which dictates the algorithm behaviour.

ironcladlou commented:
> > Thanks for that context. That seems useful for our implementation, but I think what I meant to ask was how does an end user of hypershift select which NodePools participate in autoscaling? Or are we saying that all NodePools owned by a HostedCluster are (at least in the 1st rev) subject to the rules defined in the HostedCluster Autoscaling configuration defined in this PR?
> 
> Autoscaling is already running and working in the main branch. You can opt-in for a nodePool to be autoscaled by setting `nodePool.spec.Autoscalin.min` and `nodePool.spec.Autoscalin.max`
> 
> This PR is just exposing some tweakable flags for the cluster-autoscaler binary which dictate the algorithm behaviour.

Thanks, I think this explains what I was asking. I think it would be useful to figure out how to document the scaling behavior holistically. Not sure if it belongs on HostedCluster or NodePool since it involves both.

elmiko commented:
> Thanks, I think this explains what I was asking. I think it would be useful to figure out how to document the scaling behavior holistically. Not sure if it belongs on HostedCluster or NodePool since it involves both.

+1, the docs for how to change the cluster-autoscaler algorithmic behavior, and how to utilize autoscaling in the cluster should definitely live in the same place.

JoelSpeed commented on docs/api.md:
```diff
@@ -118,8 +121,46 @@ type AddonsSpec struct {
+	ScaleDownEnabled bool `json:"enabled,omitempty"`
+
+	// Gives pods graceful termination time before scaling down
+    // default: 600 seconds
```
Indentation on all of the default lines seems to be wrong, needs fixing up across this struct

JoelSpeed commented on docs/api.md:
```diff
@@ -118,8 +121,46 @@ type AddonsSpec struct {
+	// the same instance type and the same set of labels and try
+	// to keep the respective sizes of those node groups balanced.
+    // default: false
+	BalanceSimilarNodeGroups *bool `json:"balanceSimilarNodeGroups,omitempty"`
+
+	// Enables/Disables `--ignore-daemonsets-utilization` CA feature flag. Should CA ignore DaemonSet pods when calculating resource utilization for scaling down.
+    // default: false
+	IgnoreDaemonsetsUtilization *bool `json:"ignoreDaemonsetsUtilization,omitempty"`
```
If these default to false, is there a need to have them as bool pointers? This is typically a hack to allow a nil entry to imply a true default

elmiko replied:
i'm good with that

elmiko replied:
> This is typically a hack to allow a nil entry to imply a true default

funnily enough, we use it in exactly the opposite manner in the CAO

ironcladlou replied:
I propose removing this entirely and splitting out a separate discussion about a policy enum of some kind

elmiko replied:
i chatted with Joel briefly about the pointer issue, and i'm going to leave these as pointers for the time being. the main reasoning here is that in the code we gate on a `nil` to decide if we should add the command line flag or not to the deployment spec for the autoscaler. we rely on the defaults that are part of the autoscaler binary instead of maintaining a set of defaults in the deployment code.

curious to hear if there are any other opinions about the pointers here, otherwise i will leave them as is.

JoelSpeed commented on docs/api.md:
```diff
@@ -118,8 +121,46 @@ type AddonsSpec struct {
+
+	// Enables/Disables `--skip-nodes-with-local-storage` CA feature flag. If true cluster autoscaler will never delete nodes with pods with local storage, e.g. EmptyDir or HostPath.
+    // default: true
+	SkipNodesWithLocalStorage *bool `json:"skipNodesWithLocalStorage,omitempty"`
```
Maybe controversial as this would change the name, but, what about flipping the name of this to something like `ScaleDownNodesWithLocalStorage` which would default false, then we don't need a true default here and it is, IMO, more obvious to the end user what is going on when this feature is enabled/disabled

elmiko replied:
i'm ok with changing the name if folks think it makes better sense, i have a feeling this one was just copied straight over from the command line flag (`--skip-nodes-with-local-storage`).

ironcladlou replied:
As with the other bools, I propose removing this field for now and splitting out a separate discussion about a policy enum or union type

ironcladlou commented on docs/api.md:
```diff
@@ -118,8 +121,46 @@ type AddonsSpec struct {
+	// the same instance type and the same set of labels and try
+	// to keep the respective sizes of those node groups balanced.
+    // default: false
+	BalanceSimilarNodeGroups *bool `json:"balanceSimilarNodeGroups,omitempty"`
```
I propose removing this one and splitting out a separate discussion about something like a "balancing policy" instead of a boolean

elmiko replied:
i'd like to hear more about the policy enum/union idea. that said, i'm fine to remove these for now.

elmiko commented:
updates
* removed bool fields in favor of later discussion on profiles
* fix spacing issue

ironcladlou commented:
/approve

this lgtm

csrwng commented:
/lgtm
